packages/aws-cdk-lib/aws-cloudfront-origins/lib/s3-bucket-origin.ts

```diff
@@ -0,0 +1,251 @@
+  */
+  readonly originAccessIdentity?: cloudfront.IOriginAccessIdentity;
+}
+
+/**
+ * A S3 Bucket Origin
+ */
+export abstract class S3BucketOrigin extends cloudfront.OriginBase {
+  /**
+   * Create a S3 Origin with Origin Access Control (OAC) configured
+   */
+  public static withOriginAccessControl(bucket: IBucket, props?: S3BucketOriginWithOACProps): cloudfront.IOrigin {
+    return new class extends S3BucketOrigin {
+      private originAccessControl?: cloudfront.IOriginAccessControl;
+
+      constructor() {
+        super(bucket, { ...props });
+        this.originAccessControl = props?.originAccessControl;
+      }
+
+      public bind(scope: Construct, options: cloudfront.OriginBindOptions): cloudfront.OriginBindConfig {
+        if (!this.originAccessControl) {
+          this.originAccessControl = new cloudfront.S3OriginAccessControl(scope, 'S3OriginAccessControl');
+        }
+
+        const distributionId = options.distributionId;
+        const accessLevels = new Set(props?.originAccessLevels ?? [cloudfront.AccessLevel.READ]);
+        const bucketPolicyActions = this.getBucketPolicyActions(accessLevels);
+        const bucketPolicyResult = this.grantDistributionAccessToBucket(distributionId!, bucketPolicyActions);
+
+        // Failed to update bucket policy, assume using imported bucket
+        if (!bucketPolicyResult.statementAdded) {
+          Annotations.of(scope).addWarningV2('@aws-cdk/aws-cloudfront-origins:updateImportedBucketPolicyOac',
+            'Cannot update bucket policy of an imported bucket. You will need to update the policy manually instead.\n' +
+            'See the "Setting up OAC with imported S3 buckets" section of module\'s README for more info.');
+        }
+
+        if (bucket.encryptionKey) {
+          const keyPolicyActions = this.getKeyPolicyActions(accessLevels);
+          const keyPolicyResult = this.grantDistributionAccessToKey(keyPolicyActions, bucket.encryptionKey);
+          // Failed to update key policy, assume using imported key
+          if (!keyPolicyResult.statementAdded) {
+            Annotations.of(scope).addWarningV2('@aws-cdk/aws-cloudfront-origins:updateImportedKeyPolicy',
+              'Cannot update key policy of an imported key. You may need to update the policy manually instead.');
+          }
+        }
+
+        const originBindConfig = this._bind(scope, options);
+
+        // Update configuration to set OriginControlAccessId property
+        return {
+          ...originBindConfig,
+          originProperty: {
+            ...originBindConfig.originProperty!,
+            originAccessControlId: this.originAccessControl.originAccessControlId,
+          },
+        };
+      }
+
+      private getBucketPolicyActions(accessLevels: Set<cloudfront.AccessLevel>) {
+        let actions: string[] = [];
+        for (const accessLevel of accessLevels) {
+          actions = actions.concat(BUCKET_ACTIONS[accessLevel]);
+        }
+        return actions;
```

Comment:
It's because `BUCKET_ACTIONS` (and `KEY_ACTIONS`) map the keys to an array of permissions:
```
const BUCKET_ACTIONS: Record<string, string[]> = {
  READ: ['s3:GetObject'],
  WRITE: ['s3:PutObject'],
  DELETE: ['s3:DeleteObject'],
};
```
so we need `concat` to merge the arrays

Reply:
You don't have to change it here, but just a FYI, you can also do something like `list.push(...[A,B,C])`.